### PR TITLE
Fix problem with `percentage_central` argument in `check_outliers()` with MCD method

### LINKED ISSUE #672: Investigating the high % of outliers detected with the MCD method
We have discussed how the MCD method's default threshold seems to provide high percentages of outlier detections in real-life datasets, and also as exemplified in the `mtcars` dataset. The reprex below aims to investigate this issue a bit more systematically.

<details>
<summary>
Code for compare_methods() function
</summary>

<br>

``` r
compare_methods <- function(data) {
  data <- datawizard::data_select(data, select = is.numeric)
  data[is.na(data)] <- 0
  df <- data.frame(method = c(
    "mahalanobis", "mahalanobis_robust", "mcd", "leys_mcd"))
  x <- lapply(df$method[1:3], \(x) {
    performance::check_outliers(data, method = x) |> 
      which() |> 
      length() }) |> 
    unlist()
  leys_mcd <- Routliers::outliers_mcd(data)$nb
  outliers <- c(x, leys_mcd)
  data.frame(df, outliers, percent = round(outliers / nrow(data), 2))
}
```

</br>
</details>

``` r
list.df <- tibble::lst(mtcars, iris, ToothGrowth, airquality)
lapply(list.df, compare_methods)
#> $mtcars
#>               method outliers percent
#> 1        mahalanobis        0    0.00
#> 2 mahalanobis_robust        8    0.25
#> 3                mcd       10    0.31
#> 4           leys_mcd       15    0.47
#> 
#> $iris
#>               method outliers percent
#> 1        mahalanobis        0    0.00
#> 2 mahalanobis_robust        4    0.03
#> 3                mcd        3    0.02
#> 4           leys_mcd       53    0.35
#> 
#> $ToothGrowth
#>               method outliers percent
#> 1        mahalanobis        0     0.0
#> 2 mahalanobis_robust        0     0.0
#> 3                mcd        6     0.1
#> 4           leys_mcd        0     0.0
#> 
#> $airquality
#>               method outliers percent
#> 1        mahalanobis        0    0.00
#> 2 mahalanobis_robust        1    0.01
#> 3                mcd        6    0.04
#> 4           leys_mcd       27    0.18
```

<sup>Created on 2024-01-24 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

We can see that indeed, from the `mtcars` dataset, our default threshold MCD method flags a high percentage of outliers (31%), but we can also see that Ley's MCD function's (`Routliers::outliers_mcd()`) default threshold (which is more conservative) tags a whopping 47% outliers, almost half of the sample!

However, this could be an artifact from the `mtcars` dataset, since the percentages are much lower for `iris`, `ToothGrowth`, `airquality`, ranging from 1% (our MCD) to 35% (Ley's MCD).

---

Next, I will try to get the exact percentages I was getting in my own real-life datasets later.

### PR DESCRIPTION
**Pull request by rempsyc**
Fixes #672 

`percentage_central` now works. As discussed in #672

---

Reprex:

``` r
library(performance)
alpha <- 0.1
check_outliers(mtcars, method = "mcd", 
               percentage_central = .50,
               threshold = stats::qchisq(p = 1 - alpha, df = ncol(mtcars))) |>
  which() |>
  length()
#> [1] 15

check_outliers(mtcars, method = "mcd", 
               percentage_central = .75,
               threshold = stats::qchisq(p = 1 - alpha, df = ncol(mtcars))) |>
  which() |>
  length()
#> [1] 10

check_outliers(mtcars, method = "mcd", 
               percentage_central = .25,
               threshold = stats::qchisq(p = 1 - alpha, df = ncol(mtcars))) |>
  which() |>
  length()
#> Error in MASS::cov.rob(x, quantile.used = percentage_central * nrow(x), : 'quantile' must be at least 12
```

<sup>Created on 2024-01-24 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

### CODE REVIEW

**Comment by rempsyc** [comment]
Errors appear related to `Matrix` and not to this minimal PR

**Comment by rempsyc** [comment]
@bwiernik I fixed the bug that `percentage_central` was being overwritten, but that means instead of a forced .66, it came back to a default of .50. I changed it to .75 to be consistent with Leys' 2018 recommendations, but let me know if you think it should really be .66 and not .75. If we differ from recommendations, we might have to explain this choice in the paper.

**Comment by strengejacke** [comment]
The lower the value, the more outliers detected. As I thought the method flags too many outliers, anyway, we should stick to the higher value as default?

**Comment by strengejacke** [comment]
``` r
n_outliers_MCD <- function(N, p) {
  data <- data.frame(MASS::mvrnorm(N, rep(0, p), diag(p)))
  results <- performance::check_outliers(data, method = "mcd")
  cat(sprintf("\nNumber of rows: %d, Number of columns: %d\n", nrow(data), ncol(data)))
  print(results)
  cat("\n")
}

grid <- expand.grid(
  N = c(50, 100, 150, 200, 300, 500),
  p = c(5, 10, 15, 20, 25)
)

mapply(n_outliers_MCD, N = grid$N, p = grid$p)
#> Sample size is too small resp. number of variables is too high in your
#>   data for MCD to be reliable.
#> 
#> Number of rows: 50, Number of columns: 5
#> 5 outliers detected: cases 3, 4, 5, 11, 37.
#> - Based on the following method and threshold: mcd (20.515).
#> - For variables: X1, X2, X3, X4, X5.
#> 
#> 
#> Number of rows: 100, Number of columns: 5
#> OK: No outliers detected.
#> - Based on the following method and threshold: mcd (20.515).
#> - For variables: X1, X2, X3, X4, X5
#> 
#> 
#> 
#> Number of rows: 150, Number of columns: 5
#> 1 outlier detected: case 22.
#> - Based on the following method and threshold: mcd (20.515).
#> - For variables: X1, X2, X3, X4, X5.
#> 
#> 
#> Number of rows: 200, Number of columns: 5
#> 1 outlier detected: case 189.
#> - Based on the following method and threshold: mcd (20.515).
#> - For variables: X1, X2, X3, X4, X5.
#> 
#> 
#> Number of rows: 300, Number of columns: 5
#> OK: No outliers detected.
#> - Based on the following method and threshold: mcd (20.515).
#> - For variables: X1, X2, X3, X4, X5
#> 
#> 
#> 
#> Number of rows: 500, Number of columns: 5
#> 1 outlier detected: case 465.
#> - Based on the following method and threshold: mcd (20.515).
#> - For variables: X1, X2, X3, X4, X5.
#> Sample size is too small resp. number of variables is too high in your
#>   data for MCD to be reliable.
#> 
#> Number of rows: 50, Number of columns: 10
#> 5 outliers detected: cases 20, 22, 26, 29, 42.
#> - Based on the following method and threshold: mcd (30).
#> - For variables: X1, X2, X3, X4, X5, X6, X7, X8, X9, X10.
#> Sample size is too small resp. number of variables is too high in your
#>   data for MCD to be reliable.
#> 
#> Number of rows: 100, Number of columns: 10
#> OK: No outliers detected.
#> - Based on the following method and threshold: mcd (30).
#> - For variables: X1, X2, X3, X4, X5, X6, X7, X8, X9, X10
#> 
#> 
#> 
#> Number of rows: 150, Number of columns: 10
#> 1 outlier detected: case 82.
#> - Based on the following method and threshold: mcd (30).
#> - For variables: X1, X2, X3, X4, X5, X6, X7, X8, X9, X10.
#> 
#> 
#> Number of rows: 200, Number of columns: 10
#> OK: No outliers detected.
#> - Based on the following method and threshold: mcd (30).
#> - For variables: X1, X2, X3, X4, X5, X6, X7, X8, X9, X10
#> 
#> 
#> 
#> Number of rows: 300, Number of columns: 10
#> OK: No outliers detected.
#> - Based on the following method and threshold: mcd (30).
#> - For variables: X1, X2, X3, X4, X5, X6, X7, X8, X9, X10
#> 
#> 
#> 
#> Number of rows: 500, Number of columns: 10
#> 4 outliers detected: cases 93, 180, 417, 497.
#> - Based on the following method and threshold: mcd (30).
#> - For variables: X1, X2, X3, X4, X5, X6, X7, X8, X9, X10.
#> Sample size is too small resp. number of variables is too high in your
#>   data for MCD to be reliable.
#> 
#> Number of rows: 50, Number of columns: 15
#> 12 outliers detected: cases 1, 6, 11, 14, 15, 18, 21, 24, 25, 30, 33,
#>   35.
#> - Based on the following method and threshold: mcd (40).
#> - For variables: X1, X2, X3, X4, X5, X6, X7, X8, X9, X10, X11, X12, X13,
#>   X14, X15.
#> Sample size is too small resp. number of variables is too high in your
#>   data for MCD to be reliable.
#> 
#> Number of rows: 100, Number of columns: 15
#> 4 outliers detected: cases 2, 71, 76, 96.
#> - Based on the following method and threshold: mcd (40).
#> - For variables: X1, X2, X3, X4, X5, X6, X7, X8, X9, X10, X11, X12, X13,
#>   X14, X15.
#> Sample size is too small resp. number of variables is too high in your
#>   data for MCD to be reliable.
#> 
#> Number of rows: 150, Number of columns: 15
#> 1 outlier detected: case 29.
#> - Based on the following method and threshold: mcd (40).
#> - For variables: X1, X2, X3, X4, X5, X6, X7, X8, X9, X10, X11, X12, X13,
#>   X14, X15.
#> 
#> 
#> Number of rows: 200, Number of columns: 15
#> 5 outliers detected: cases 24, 48, 54, 98, 146.
#> - Based on the following method and threshold: mcd (40).
#> - For variables: X1, X2, X3, X4, X5, X6, X7, X8, X9, X10, X11, X12, X13,
#>   X14, X15.
#> 
#> 
#> Number of rows: 300, Number of columns: 15
#> 5 outliers detected: cases 23, 30, 175, 230, 240.
#> - Based on the following method and threshold: mcd (40).
#> - For variables: X1, X2, X3, X4, X5, X6, X7, X8, X9, X10, X11, X12, X13,
#>   X14, X15.
#> 
#> 
#> Number of rows: 500, Number of columns: 15
#> 2 outliers detected: cases 53, 314.
#> - Based on the following method and threshold: mcd (40).
#> - For variables: X1, X2, X3, X4, X5, X6, X7, X8, X9, X10, X11, X12, X13,
#>   X14, X15.
#> Sample size is too small resp. number of variables is too high in your
#>   data for MCD to be reliable.
#> 
#> Number of rows: 50, Number of columns: 20
#> 12 outliers detected: cases 2, 3, 5, 6, 13, 19, 23, 27, 29, 33, 36, 45.
#> - Based on the following method and threshold: mcd (50).
#> - For variables: X1, X2, X3, X4, X5, X6, X7, X8, X9, X10, X11, X12, X13,
#>   X14, X15, X16, X17, X18, X19, X20.
#> Sample size is too small resp. number of variables is too high in your
#>   data for MCD to be reliable.
#> 
#> Number of rows: 100, Number of columns: 20
#> 19 outliers detected: cases 4, 5, 8, 10, 29, 45, 50, 51, 55, 60, 66, 77,
#>   78, 81, 85, 90, 93, 94, 96.
#> - Based on the following method and threshold: mcd (50).
#> - For variables: X1, X2, X3, X4, X5, X6, X7, X8, X9, X10, X11, X12, X13,
#>   X14, X15, X16, X17, X18, X19, X20.
#> Sample size is too small resp. number of variables is too high in your
#>   data for MCD to be reliable.
#> 
#> Number of rows: 150, Number of columns: 20
#> 6 outliers detected: cases 27, 37, 82, 119, 127, 130.
#> - Based on the following method and threshold: mcd (50).
#> - For variables: X1, X2, X3, X4, X5, X6, X7, X8, X9, X10, X11, X12, X13,
#>   X14, X15, X16, X17, X18, X19, X20.
#> Sample size is too small resp. number of variables is too high in your
#>   data for MCD to be reliable.
#> 
#> Number of rows: 200, Number of columns: 20
#> 3 outliers detected: cases 42, 145, 197.
#> - Based on the following method and threshold: mcd (50).
#> - For variables: X1, X2, X3, X4, X5, X6, X7, X8, X9, X10, X11, X12, X13,
#>   X14, X15, X16, X17, X18, X19, X20.
#> 
#> 
#> Number of rows: 300, Number of columns: 20
#> 3 outliers detected: cases 4, 146, 261.
#> - Based on the following method and threshold: mcd (50).
#> - For variables: X1, X2, X3, X4, X5, X6, X7, X8, X9, X10, X11, X12, X13,
#>   X14, X15, X16, X17, X18, X19, X20.
#> 
#> 
#> Number of rows: 500, Number of columns: 20
#> 3 outliers detected: cases 226, 241, 424.
#> - Based on the following method and threshold: mcd (50).
#> - For variables: X1, X2, X3, X4, X5, X6, X7, X8, X9, X10, X11, X12, X13,
#>   X14, X15, X16, X17, X18, X19, X20.
#> Sample size is too small resp. number of variables is too high in your
#>   data for MCD to be reliable.
#> 
#> Number of rows: 50, Number of columns: 25
#> 11 outliers detected: cases 1, 2, 7, 20, 24, 31, 33, 38, 41, 45, 47.
#> - Based on the following method and threshold: mcd (52.62).
#> - For variables: X1, X2, X3, X4, X5, X6, X7, X8, X9, X10, X11, X12, X13,
#>   X14, X15, X16, X17, X18, X19, X20, X21, X22, X23, X24, X25.
#> Sample size is too small resp. number of variables is too high in your
#>   data for MCD to be reliable.
#> 
#> Number of rows: 100, Number of columns: 25
#> 21 outliers detected: cases 1, 3, 8, 9, 10, 15, 19, 28, 32, 40, 49, 56,
#>   60, 62, 63, 73, 76, 80, 92, 94, 99.
#> - Based on the following method and threshold: mcd (52.62).
#> - For variables: X1, X2, X3, X4, X5, X6, X7, X8, X9, X10, X11, X12, X13,
#>   X14, X15, X16, X17, X18, X19, X20, X21, X22, X23, X24, X25.
#> Sample size is too small resp. number of variables is too high in your
#>   data for MCD to be reliable.
#> 
#> Number of rows: 150, Number of columns: 25
#> 12 outliers detected: cases 28, 29, 78, 83, 86, 101, 103, 115, 128, 136,
#>   146, 148.
#> - Based on the following method and threshold: mcd (52.62).
#> - For variables: X1, X2, X3, X4, X5, X6, X7, X8, X9, X10, X11, X12, X13,
#>   X14, X15, X16, X17, X18, X19, X20, X21, X22, X23, X24, X25.
#> Sample size is too small resp. number of variables is too high in your
#>   data for MCD to be reliable.
#> 
#> Number of rows: 200, Number of columns: 25
#> 6 outliers detected: cases 12, 40, 51, 90, 94, 155.
#> - Based on the following method and threshold: mcd (52.62).
#> - For variables: X1, X2, X3, X4, X5, X6, X7, X8, X9, X10, X11, X12, X13,
#>   X14, X15, X16, X17, X18, X19, X20, X21, X22, X23, X24, X25.
#> 
#> 
#> Number of rows: 300, Number of columns: 25
#> 3 outliers detected: cases 15, 161, 247.
#> - Based on the following method and threshold: mcd (52.62).
#> - For variables: X1, X2, X3, X4, X5, X6, X7, X8, X9, X10, X11, X12, X13,
#>   X14, X15, X16, X17, X18, X19, X20, X21, X22, X23, X24, X25.
#> 
#> 
#> Number of rows: 500, Number of columns: 25
#> 6 outliers detected: cases 85, 171, 259, 313, 351, 390.
#> - Based on the following method and threshold: mcd (52.62).
#> - For variables: X1, X2, X3, X4, X5, X6, X7, X8, X9, X10, X11, X12, X13,
#>   X14, X15, X16, X17, X18, X19, X20, X21, X22, X23, X24, X25.
```

<sup>Created on 2024-02-04 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

File: R/check_outliers.R [hunk 1]
```diff
@@ -1726,14 +1730,31 @@ check_outliers.metabin <- check_outliers.metagen
...
+  # See #672: This does seem to be a function of the N/p (N = sample size; p =
+  # number of parameters) ratio. When it is larger than 10, the % of outliers
+  # flagged is okay (in well behaved data). This makes sense: the MCD looks at
+  # the cov matrix of subsamples of the data - with high dimensional data, small
+  # samples sizes will give highly variable cov matrices, as so the "smallest"
+  # one will probably miss-represent the data.
+
+  if ((nrow(x) / ncol(x)) <= 10 && isTRUE(verbose)) {
+    insight::format_alert(
+      "Sample size is too small resp. number of variables is too high in your data for MCD to be reliable.",
```
Comment (rempsyc) [inline R/check_outliers.R]: What is the meaning of "resp."? If it is an abbreviation, I think I would spell it out for clarity

Comment (strengejacke) [reply R/check_outliers.R]: respectively

**Comment by mattansb** [comment]
@strengejacke I think this should be a proper warning.

**Comment by strengejacke** [comment]
@rempsyc You may want to add one or two sentences, in a similar fashion like we did in our paper, to the vignette. But that's not urgent, maybe in a different PR.

File: R/check_outliers.R (hunk 1 above)
Comment (rempsyc) [reply R/check_outliers.R]: I reworded the message a bit for clarity, I hope it's OK

**Comment by strengejacke** [comment]
Thanks! Remaining failing tests should not be related to this PR.

**Pull request merged**